include/revng/AutoEnforcer/AutoEnforcerLibraryRegistry.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+
+#include "revng/AutoEnforcer/PipelineLoader.h"
+
+namespace AutoEnforcer {
+
+class AutoEnforcerLibraryRegistry {
+public:
+  AutoEnforcerLibraryRegistry() { getInstances().push_back(this); }
+
+  static void registerAllContainersAndEnforcers(PipelineLoader &Loader) {
+    for (const auto &Reg : getInstances())
+      Reg->registerContainersAndEnforcers(Loader);
+  }
+
+  static KindsRegisty registerAllKinds() {
+    KindsRegisty Registry;
+    for (const auto &Reg : getInstances())
+      Reg->registerKinds(Registry);
+
+    for (auto &Kind : Registry) {
+      Kind.getRootAncestor()->assign();
+      Kind.Granularity->getRootAncestor()->assign();
+    }
+    return Registry;
+  }
+
+  virtual ~AutoEnforcerLibraryRegistry(){};
+
+  virtual void registerContainersAndEnforcers(PipelineLoader &Loader) = 0;
+  virtual void registerKinds(KindsRegisty &KindDictionary) = 0;
+
+private:
+  static llvm::SmallVector<AutoEnforcerLibraryRegistry *, 3> &getInstances();
```

Comment:
I have the general feeling that having odd-sized `SmallVectors` is weird. Alignment, cache lines...
What do you think?

Reply:
i no longer remember why i started using 3, i think it was because i thought something like "probably one keeps 3 floats toghether to rappresents coordinate and 4 is unlikelly" but yeah i can change them to 4